app/src/main/java/com/kickstarter/viewmodels/ChangePasswordViewModel.kt

```diff
@@ -1,165 +1,58 @@
-
-        private val changePasswordClicked = PublishSubject.create<Unit>()
-        private val confirmPassword = PublishSubject.create<String>()
-        private val currentPassword = PublishSubject.create<String>()
-        private val newPassword = PublishSubject.create<String>()
-
-        private val error = BehaviorSubject.create<String>()
-        private val progressBarIsVisible = BehaviorSubject.create<Boolean>()
-        private val success = BehaviorSubject.create<String>()
-
-        val inputs: Inputs = this
-        val outputs: Outputs = this
-
-        private val apolloClient = requireNotNull(this.environment.apolloClientV2())
-        private val analytics = this.environment.analytics()
-
-        private val disposables = CompositeDisposable()
-
-        init {
-
-            val changePassword = Observable.combineLatest(
-                this.currentPassword.startWith(""),
-                this.newPassword.startWith(""),
-                this.confirmPassword.startWith("")
-            ) { current, new, confirm -> ChangePassword(current, new, confirm) }
-
-            val changePasswordNotification = changePassword
-                .compose(takeWhenV2(this.changePasswordClicked))
-                .switchMap { cp -> submit(cp).materialize() }
-                .share()
-
-            changePasswordNotification
-                .compose(errorsV2())
-                .subscribe { error ->
-                    error?.localizedMessage?.let { message ->
-                        this.error.onNext(message)
-                    }
-                }
-                .addToDisposable(disposables)
-
-            changePasswordNotification
-                .compose(valuesV2())
-                .map { it.updateUserAccount()?.user()?.email() }
-                .subscribe { email ->
-                    this.analytics?.reset()
-                    email?.let {
-                        this.success.onNext(it)
-                    }
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+
+const val GENERIC_ERROR = "GENERIC"
+
+class ChangePasswordViewModel(val environment: Environment) : ViewModel() {
+
+    private val mutableError = MutableStateFlow("")
+    val error: Flow<String> get() = mutableError
+
+    private val mutableSuccess = MutableStateFlow("")
+    val success: Flow<String> get() = mutableSuccess
+
+    private val mutableIsLoading = MutableStateFlow(false)
+    val isLoading: Flow<Boolean> get() = mutableIsLoading
+
+    private val apolloClient = requireNotNull(this.environment.apolloClientV2())
+    private val analytics = this.environment.analytics()
+
+    fun updatePassword(oldPassword: String, newPassword: String) {
+        viewModelScope.launch {
+            mutableIsLoading.emit(true)
+            val response =
+                apolloClient.updateUserPasswordNew(oldPassword, newPassword, newPassword)
```

Comment:
Could we explore something? We are re-writting with coroutines the Networking layer as well (KSApolloClientV2), which is perfectly correct, but in order to reduce the amount of layers and code we are changing could we explore as well the usage of [as-flow](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-rx2/kotlinx.coroutines.rx2/as-flow.html) method ? Or any of the other utilities present on https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-rx2/kotlinx.coroutines.rx2/as-flow.html 

I would like to figure out if to adopt coroutines is mandatory to re-write the network layer, or can it can be done in a more reduced scope, just on the UI layer with (Actvity/Frgments/VM's).

If the conclusion is that we need to rewrite the network layer I'm perfectly ok with it (long term it's what will happen anyway), but want to know all the available options here.

Reply:
I can play around with that for sure.  I think eventually we would re-write it so this is my all in approach while still trying tnot to change very much when it comes to the network layer (you can see the call to updateUserPasswordNew doesnt change very much from the non new version other than the await() call which does the response vs error handling for us).  My guess is that while we can use asFlow the amount of effort to update it to use await (or newer calls in the newer library versions) its about the same.

Reply:
want to make sure we do not block any thread by using `await` (it should not if we use it as a promise) but wanna explore all the options, if there is a bridge RX - coroutines wanna see how it behaves before doing a full rewrite. The main reason at the present moment to explore the bridge is to not juggle with so many migrations/adoptions as we currently are. But completely agree on the point that eventually there will be a full re-writte for the networking layer, but was hopping to delay that re-writting to the moment were we can adopt `repository` pattern and possibly get rid of the V1 API, but let's discuss further on the Android sync, it's quite a fascinating topic ❤️